File: baekjoon/18111.py
import sys
from collections import Counter

temp,lands=[],[]
n,m,b=map(int,sys.stdin.readline().split())
count,ispossible=1,True
for _ in range(n):
    temp.append(list(map(int,sys.stdin.readline().split())))
for i in range(n):
    for j in range(m):
        lands.append(temp[i][j])
lands.sort(reverse=True)
cnt=Counter(lands)
print(lands)
while True:
    target,time=cnt.most_common()[1][count],0
    for land in lands:
        if land>target:
            b+=land-target
            time+=1
        elif land<target:
            if b>0:
                b-=target-land
                time+=2
            else:
                ispossible=False
                break
    if ispossible:
        break
    count+=1

print(time)
#cnt부분 다시

File: baekjoon/18111.cpp
#include <bits/stdc++.h>
#include <iostream>
using namespace std;

int main(){
    int n, m, b, temp, time=1000000000, top_level, maxh;
    cin >> n >> m >> b;
    vector<int> lands;
    
    for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
            cin >> temp;
            maxh+=temp;
            lands.push_back(temp);
        }
    }
    sort(lands.begin(), lands.end());
    maxh+=b; maxh=maxh/(n*m);
    for(int target=0; target<=maxh; target++){
        int ctime=0, hand=b; bool ispossible=true;

        for(int land: lands){
            if(land>target){
                hand+=land-target;
                ctime+=2*(land-target);
            }else if(land<target){
                hand-=target-land;
                ctime+=target-land;
            }
            if(ctime>time){
                ispossible=false;
                break;
            }
        }
        if(hand<0) ispossible=false;
        if(!ispossible) continue;
        if (ctime<time){
            top_level=target;
            time=ctime;
        }
        if(ctime==time && target>top_level){
            top_level=target;
        }
    }

    cout << time << " " << top_level;

    return 0;
}

/// lands 원소만 생각하는게 아닌 다른 경우도 찾아보기 중간값에서도 정답이 나올수있음